### PR TITLE
Convert syslog service into Rust

### PR DESCRIPTION
**Pull request by jgrund**
We have a syslog service (that reads from the journal) that is written in Python on both the manager
and agent.

We can convert it to Rust by using the `systemd-journal-gatewayd.service` which will allow for us to pull data on each tick without needing any native dependencies.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2062)

### CODE REVIEW

File: iml-agent/src/daemon_plugins/journal.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    let client = reqwest::Client::new();
+
+    let range = if let Some(x) = cursor {
+        format!("entries={}:1:1000", x)
+    } else {
+        "entries=:-1000:".into()
+    };
+
+    let s = client
+        .get("http://localhost:19531/entries")
```
Comment (ip1981) [inline iml-agent/src/daemon_plugins/journal.rs:62]: Maybe it's worth a command line option for the port at least.

File: iml-services/iml-journal/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+fn client_connection_parser(msg: &str) -> Option<(i32, String)> {
+    lazy_static! {
+        static ref TARGET_END: Regex = Regex::new(r":\s+connection from").unwrap();
+    }
+
+    // get the client NID out of the string
+    let nid_start = msg.find("@")? + 1;
+    let nid_len = msg[nid_start..].find(" ")?;
+    // and the UUID
+    let uuid_start = msg.find(" from ")? + 6;
```
Comment (ip1981) [inline iml-services/iml-journal/src/lib.rs:132]: perhaps, use regex for the entire thing?

File: iml-services/iml-journal/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+            host_id,
+        )
+        .err_into()
+        .boxed();
+    };
+
+    future::ok(()).boxed()
+}
+
+fn get_item_after<'a>(s: &'a str, after: &str) -> Option<&'a str> {
```
Comment (ip1981) [inline iml-services/iml-journal/src/lib.rs:269]: I think [strip_prefix](https://doc.rust-lang.org/beta/std/primitive.str.html#method.strip_prefix) is close to this.

File: iml-services/iml-journal/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use insta::assert_debug_snapshot;
+
+    #[test]
+    fn test_get_message_class() {
```
Comment (ip1981) [inline iml-services/iml-journal/src/lib.rs:305]: For the record, I believe this is a good test :)

File: iml-services/iml-journal/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    let mut num_remove: i64 = num_rows - *DBLOG_LW;
+
+    while num_remove > 0 {
+        let xs = sqlx::query!(
+            r#"
+                DELETE FROM chroma_core_logmessage
+                WHERE id in ( 
+                    SELECT id FROM chroma_core_logmessage ORDER BY id LIMIT $1
+                )
+                RETURNING id"#,
```
Comment (ip1981) [inline iml-services/iml-journal/src/main.rs:36]: Much better than diesel :)

File: iml-agent/src/daemon_plugins/journal.rs (hunk 1 above)
Comment (jgrund) [reply iml-agent/src/daemon_plugins/journal.rs:62]: Sure. I'll add an environment var for it.

File: iml-services/iml-journal/README.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# IML Journal
+
+This crate consumes incomping journal messages from the cluster and persists them to the database.
```
Comment (jgrund) [inline iml-services/iml-journal/README.md:3]:
```suggestion
This crate consumes incoming journal messages from the cluster and persists them to the database.
```

File: iml-services/iml-journal/src/lib.rs (hunk 2 above)
Comment (jgrund) [reply iml-services/iml-journal/src/lib.rs:132]: This was pretty much a straight port from the Python equivalent (including tests). I'd rather use a parser combinator here honestly, but it's worth waiting till we have more items to match over.

**Comment by jgrund** [comment]
___
*[iml-services/iml-ntp/src/main.rs, line 27 at r1](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2062#-MCBatOL5dBZykWji4mO:-MCCVr_51bUsgvR2eyAD:b-qlprdu) ([raw file](https://github.com/whamcloud/integrated-manager-for-lustre/blob/be71f6a41883b0e8d645d788602808135f7b1c6d/iml-services/iml-ntp/src/main.rs#L27)):*
<details><summary><i>Previously, mkpankov (Michael Pankov) wrote…</i></summary><blockquote>

Maybe pack this `query_as` to a function as it was before?
</blockquote></details>

Eventually it should live in it's own module within iml-postgres, but I'm going to keep it like this for now.

**Comment by jgrund** [comment]
___
*[iml-agent/src/daemon_plugins/journal.rs, line 94 at r1](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2062#-MCBUxhaDVQDEalTVjs2:-MCChM0x-nVpsO8x2P7P:b-pjuo8u) ([raw file](https://github.com/whamcloud/integrated-manager-for-lustre/blob/be71f6a41883b0e8d645d788602808135f7b1c6d/iml-agent/src/daemon_plugins/journal.rs#L94)):*
<details><summary><i>Previously, mkpankov (Michael Pankov) wrote…</i></summary><blockquote>

What is `3`?
</blockquote></details>

https://en.wikipedia.org/wiki/Syslog

Surprisingly some items don't have a syslog_facility to this defaults it

File: iml-services/iml-journal/src/main.rs (hunk 5 above)
Comment (jgrund) [reply iml-services/iml-journal/src/main.rs:36]: Agreed

File: iml-services/iml-journal/src/lib.rs (hunk 3 above)
Comment (jgrund) [reply iml-services/iml-journal/src/lib.rs:269]: Not worth fixing. Once there are more matches, plan is to switch to parser-combinator a-la https://docs.rs/combine/4.2.1/combine/.

File: iml-postgres/src/lib.rs [hunk 7]
```diff
@@ -17,6 +21,35 @@ pub use tokio_postgres::{
 };
 use tokio_postgres::{tls::NoTlsStream, Connection, NoTls, Socket};
 
+pub async fn get_db_pool() -> Result<PgPool, sqlx::Error> {
+    let mut opts = PgConnectOptions::default().username(&iml_manager_env::get_db_user());
+
+    opts = if let Some(x) = iml_manager_env::get_db_host() {
```
Comment (johnsonw) [inline iml-postgres/src/lib.rs:27]: Would it be cleaner to do something like:
```
opts = iml_manager_env::get_db_host().map_or(opts, |x| opts.host(&x));
opts = iml_manager_env::get_db_name().map_or(opts, |x| opts.database(&x));
opts = iml_manager_env::get_db_password().map_or(opts, |x| opts.password(&x));
```
?

File: iml-services/iml-journal/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+        .await?;
+
+        let (id, msg) = match row {
+            Some(Row { id, message: Some(message) }) => (id, format!("{} {}", message, msg)),
+            Some(Row { message:None, ..}) | None => return Ok(()),
+        };
+
+        sqlx::query!(
+            r#"
+            UPDATE chroma_core_alertstate
```
Comment (johnsonw) [inline iml-services/iml-journal/src/lib.rs:189]: It probably doesn't matter but do we want the whitespace?

File: iml-postgres/src/lib.rs (hunk 7 above)
Comment (jgrund) [reply iml-postgres/src/lib.rs:27]: Subjective I suppose

**Comment by jgrund** [comment]
Pass in https://build.whamcloud.com/job/manager-for-lustre/1269/

**Pull request merged**